src/pull-request-core.ts

```diff
@@ -0,0 +1,234 @@
+import Octokit, { GitUpdateRefResponse } from '@octokit/rest';
+
+import {
+    PullRequest,
+    PullRequestWithChangesetCommit,
+    SourceFile
+} from './types';
+
+export function issuePullRequestWithToken(
+    { pullRequest, changesetCommit }: PullRequestWithChangesetCommit,
+    githubAccessToken: string
+): Promise<Octokit.PullsCreateResponse> {
+    const octokit = new Octokit({
+        auth: githubAccessToken
+    });
+
+    return Promise.resolve(issuePullRequest({ pullRequest, changesetCommit }));
+
+    function issuePullRequest({
+        pullRequest: pr,
+        changesetCommit: cs
+    }: PullRequestWithChangesetCommit): Promise<Octokit.PullsCreateResponse> {
+        return Promise.resolve(
```

Comment:
In general: assume I know basically nothing about TypeScript preferred conventions. I went about this by chaining promises and then returning them. I'm aware of `async`/`await` but don't know the hard-and-fast rules for using them instead of promises, so left it this way to promote discussion

Reply:
You should always prefer `async`/`await`.
- Sequential code is easier to read and uses the same constructs as sync code (i.e. try/catch)
- You are protected against sync exceptions too and don't need to handle one with `try`/`catch` and the other with `.catch()`
- You don't end up nesting functions further and further to keep previous results in scope

Reply:
Btw, VS Code/TypeScript has an automated refactor I think (through the 💡 when you put a cursor) to convert Promise to async/await code.

Reply:
That makes loads of sense now, I see the connection between `async`/`await` vs chained promises coming from the functional world I'm used to :)

Yeah I'm doing all this in VS Code. I think I had everything set up right but apparently VS Code refuses to give me this refactor option. It gives me lots of suggestions for other things so ¯\_(ツ)_/¯. NBD